### ISSUE TITLE
Repo split

### ISSUE THREAD

**Issue by hu55a1n1** [informalsystems/hermes#2639]
## Summary
It has been proposed that we split the ibc-rs monorepo into two repos - one for the proto & module crates and the other for the relayer crate (and its associated crates). 
Additionally, the relayer code must not depend on the `ibc` crate to have clear boundaries/separation and allow the modules to evolve independently. 

## Problem Statement
The monorepo approach is restrictive and becoming difficult to maintain. (See https://github.com/informalsystems/ibc-rs/discussions/1437#discussioncomment-1458578)
Moreover, having the relayer crate depend on the ibc crate makes it difficult to modify one without touching the other, and have resulted in code bloat and unclear boundaries - for e.g. the modules code contains relayer specific code. 

## Proposal
Using the ideas described in https://github.com/informalsystems/ibc-rs/issues/2541 and https://github.com/informalsystems/ibc-rs/discussions/1437, the repo hosting the relayer crates (the `relayer` repo) would look like so -> 

```
<relayer>
├── CHANGELOG.md
├── CONTRIBUTING.md
├── Cargo.lock
├── Cargo.toml
│
├── crates
│	├── ibc-relayer
│	├── ibc-relayer-cli
│	├── ibc-relayer-rest
│	├── ibc-telemetry
│	└── ibc-chain-registry 
│
├── etc...
```
And the `modules` repo would look like so ->

```
<modules>
├── CHANGELOG.md
├── CONTRIBUTING.md
├── Cargo.toml
│
├── crates
│	├── ibc
│	└── proto
│		├── ibc-proto
│		└── ibc-proto-compiler 
│
├── etc...
```

## Acceptance Criteria
The ibc-rs repo is split into two repos as proposed above, ideally preserving history.

### TODOs
- [ ] Rename the default branches of both repos from `master` to `main`
- [ ] Split `gm` out of the repository

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by mzabaluev** [comment]
I think the history should be preserved unchanged in both repositories after the split (without making it an official fork on GitHub). The repo size of 39 MB is not something I would worry about.

This way, anybody with an outstanding branch can rebase or merge against either or both the new repositories.

**Comment by mzabaluev** [comment]
I would also take exception to `ibc-proto-compiler` and move it under `tools`, like in `tendermint-rs`.

**Comment by hu55a1n1** [comment]
> I think the history should be preserved unchanged in both repositories after the split (without making it an official fork on GitHub).

With my current proposal, the history will be preserved unchanged for the relayer repo. The module repo will also preserve the history but only for the modules, proto & proto-compiler crates, because we're using `git-filter-repo`. I like to use `git-filter-repo` for the modules repo because it reduces the repo size to ~8.5MB, the history is cleaner and a rebase should still be possible in most cases. Here's what the resulting module repo history would look like -> https://github.com/hu55a1n1/ibc-rust/commits/master

> I would also take exception to ibc-proto-compiler and move it under tools, like in tendermint-rs.

Do you mean we move the `ibc-proto-compiler` to `crates/tools/ibc-proto-compiler` or `tools/ibc-proto-compiler`?

**Comment by mzabaluev** [comment]
> Do you mean we move the `ibc-proto-compiler` to `crates/tools/ibc-proto-compiler` or `tools/ibc-proto-compiler`?

To `tools/ibc-proto-compiler`. Also, all library crates would be on the same level under `crates`.
The logic here is that all production code is hosted under `crates`, and everything auxiliary that is used to build or maintain the production code goes under `tools`.

In `tendermint-rs`, `tools` is also an independent Cargo workspace with its own `Cargo.lock` that is version-controlled; I think it makes sense particularly for `ibc-proto-compiler`, which needs a reproducibility anchor for the code that is emitted for `ibc-proto`.

**Comment by plafer** [comment]
Two minor thoughts:
1. I think we should use `ibc-rs` for the modules, and have a statement in the README that redirects to the hermes repo and explains the transition that happened. In the long run, we'll be happy to have "the correct name" for the repo
2. Wouldn't it be better to have `proto` be its own repository? I find it cleaner for people who look for "ibc-rs" to just fall right on the "handlers repo" rather than a monorepo that also includes a proto repo.

**Comment by mzabaluev** [comment]
> Wouldn't it be better to have proto be its own repository?

I'm generally against splitting crates that are tightly coupled, like handlers and ibc-proto need to be. This leads to increased developer logistics when changes are introduced. Having to synchronize a change across two repositories (ibc-rs and hermes) is already a burden, but at least there are good reasons for this split, including the tentative idea to get rid of code reuse between modules and hermes.

**Comment by hu55a1n1** [comment]
As discussed with Philippe and Romain, we decided to keep the proto crates in a different repository for now since both the relayer and modules depend on the raw types with each repo having it's own version(s) of the domain types. 

We seem to have consensus on 'hermes' as the relayer repo's name, but we're still undecided on the modules repo's name as that topic seems more nuanced than expected. `ibc-rs` seems like the favourite (followed by `ibc-rust`) but we first need to answer the following questions before we can take a decision on this ->
* What happens when someone pushes to a remote that has been renamed? 
* What if the remote's old name has been reused?
* Are there long-lived forks/branches of hermes? (looks like there are ~5 projects that have recently updated forks)
* Name of forks wont change, so could this be a problem?
* What about links to ibc-rs in external repos?

**Comment by adizere** [comment]
An alternative is to create a repo `cosmos/ibc-rs` to host the modules in Rust. Context I can provide here to inform the decision is that this alternative would work towards incentivizing external contribution to modules, especially since that part of our work is more outside oriented. Drawback is we lose the "Informal brand".

I think hosting the Hermes repo under `informalsystems/hermes` (renamed and redirected automatically from `informalsystems/ibc-rs`) makes a lot of sense and I have no input there.

**Comment by romac** [comment]
Yeah let's re-use that repo 👍🏻

**Comment by romac** [comment]
While we're at it, we should use this opportunity to rename the default branch from `master` to `main` as per the new Git conventions: https://github.com/github/renaming

**Comment by greg-szabo** [comment]
I would use this event to break out `gm` from under IBC so it can start becoming its own tool.

**Comment by adizere** [comment]
> I would use this event to break out `gm` from under IBC so it can start becoming its own tool.

Might also be a good idea to rename from `gm` = "gaia manager" to something more broad, e.g., "local ibc/interchain testnet manager" or something along those lines. Not sure what acronym would that yield, `lite` comes to mind.

**Comment by harveenSingh** [comment]
Good idea but when we use the tool it's `gm` for example: `gm start`. Are you suggesting to change that as well? 
Adding @greg-szabo

**Comment by adizere** [comment]
> Good idea but when we use the tool it's `gm` for example: `gm start`. Are you suggesting to change that as well?
> Adding @greg-szabo

Yes, basically rename the binary. @harveenSingh can you please create a separate issue in the gm repo to track renaming `gm` and continue this discussion? Thanks!
Curious what @greg-szabo thinks about renaming `gm`.



I will close this as the repo split is done.
